ReadAsync and WriteAsync implementation for .NET standard 1.5

meinsiedler opened the pull request:
Hi!

This PR resolves the issue described in #114. As discussed there, .NET standard doesn't support the `Begin|EndRead` and `Begin|EndWrite` methods and the current implementation of SerialPortStream didn't implement the `ReadAsync` and `WriteAsync` methods from the `Stream` base class.

With this PR, the two methods are now implemented. The two new methods simply delegate to the existing `BeginRead` and `BeginWrite` methods using the `Task.FromAsync` method. Since `AsyncResult` isn't available any more too, we can use a `Task` instead as described [here](https://stackoverflow.com/a/55516918/1384848) (`Task` implements `IAsyncResult`).

With the approach of re-using the existing `Begin|EndRead` and `Begin|EndWrite` methods we can leverage the well tested and established code without introducing much more code to maintain.

Unfortunately, the Asynchronous Programming Model (APM) (using `Begin|EndRead` and `Begin|EndWrite` methods) didn't consider cancellation, so we cannot support the `CancallationToken` with the new `ReadAsync` and `WriteAsync` methods easily. The passsed `cancellationToken` will simply be ignored. We hope that's OK for now. Otherwise we would need to rewrite the entire read/write process using "real" asynchronicity and with "real" cancellation behavior. Since this would mean a lot more effort and more further code to maintain in future for little use, we think it's better to ignore the cancellation behavior in that case.

Regarding testing: We haven't added any more unit tests to your solution, since we saw that no .NET standard 1.5 tests were present. For testing our changes locally, we added my unit test from [my test project](https://github.com/meinsiedler/SerialPortStreamTest/blob/main/SerialPortStreamTest.Tests/SerialPortStreamTest.cs) directly, but didn't commit it to your repo. The tests now succeed with .NET Core 3.1 too.

We'd be very happy if our PR for fixing the issues with .NET Core lands in a new NuGet version of your project.

Best regards.

jcurl commented on code/SerialPortStream.cs:
```diff
@@ -699,9 +700,26 @@ private int InternalRead(byte[] buffer, int offset, int count)
-#if !NETSTANDARD15
-        private delegate int ReadDelegate(byte[] buffer, int offset, int count);
-
+#if NETSTANDARD15
```
There's no reason this needs to be NETSTANDARD only, it should work just as well for .NET 4.5 and later. This would allow unit testing also for .NET 4.5. Honestly, the only reason I didn't have test cases for .NET Standard is because they were there for .NET FX and the differences were trivial.

Any chances to make this PR for both with NETSTANDARD15 || NET45, and then add your Unit Test cases at least for NET45, which should be the same/similar?

meinsiedler replied:
> There's no reason this needs to be NETSTANDARD only, it should work just as well for .NET 4.5 and later

Correct. I will adapt it to use it for both, NETSTANDARD and .NET 4.5 or later.

>  and then add your Unit Test cases at least for NET45, which should be the same/similar

Yes, I'll add the test for .NET 4.5. My unit test has worked for .NET 4.5 before the change in this PR too, so the test wouldn't prove that it works for NETSTANDARD, but at least it shows that .NET 4.5 still works with these new changes.

jcurl replied:
You're correct, you won't find the problem you found with .NET Standard by testing with .NET 4.5, but I hope there'll be at least some code coverage showing it works. Eventually, I'll get around to modernizing this a bit (upgrading to NUnit 3.x) and try to figure out how to get the unit tests working under dotnet core. That'll be later, and the unit tests will help none-the-less.

jcurl reviewed (changes requested):
Is it possible to keep the same coding style as the original? No "this.", braces on the same line.

jcurl commented:
I really appreciate your effort in making the changes. Had a quick look over it and only some minor details.

meinsiedler commented:
> I really appreciate your effort in making the changes. Had a quick look over it and only some minor details.

I also really appreciate your quick responses and your effort you make by providing useful information and your feedback. 😊

I'll update the PR with the requested changes maybe tomorrow or next week at latest.

meinsiedler commented:
Hi,

I enabled `ReadAsync` and `WriteAsync` for .NET 4.5 too. Also, I added a unit test for .NET 4.5 only and fixed some code formatting issues (braces, `#region` indentation).

jcurl commented on code/SerialPortStream.cs:
```diff
@@ -1205,19 +1255,54 @@ public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, As
+
+        private IAsyncResult InternalBeginWrite(byte[] buffer, int offset, int count, AsyncCallback callback, object state)
+        {
+            this.WriteCheck(buffer, offset, count);
```
For me as a reminder: I'll check if this is better in the public API.

jcurl reviewed (approved):
Sorry, I just saw that the new changes were pending my review.

As soon as I get some time, I'll check out the changes, rebase on what I've got at the same time (I've got a couple of other commits), and squash, cleaning it up into a single commit. I'll want to do some testing on Windows first with my USB/SER devices, and probably a quick check on Linux won't hurt. I'll probably want to experiment with some changes before I push.

Please be patient before I push it out, it might be 1-2 weeks depending on my work schedule.

jcurl commented:
Integrated with release v2.3.0 and published on NuGet.
